### PR TITLE
add sliders to settings

### PR DESCRIPTION
**Pull request by kotek900**
A compact, short information about PR for easier understanding:

- Sliders for settings
- More user friendly that typing in numbers

## To do

This PR is a ~~Work in Progress~~ / Ready for Review.

- [ ] add button to turn slider into a regular text box
- [ ] add indicators to sliders

## How to test

image

### CODE REVIEW

File: builtin/mainmenu/settings/components.lua [hunk 1]
```diff
@@ -126,6 +126,40 @@ end, is_valid_number)
...
+
+	return {
+		info_text = setting.comment,
+		setting = setting,
+
+		get_formspec = function(self, avail_w)
+			local value = core.settings:get(setting.name) or setting.default
+			local scrollbar_value = value/(setting.max - setting.min)*steps - setting.min
+
+			self.resettable = true
```
Comment (y5nw) [inline builtin/mainmenu/settings/components.lua:140]:
```suggestion
			self.resettable = core.settings:has(setting.name)
```

File: builtin/mainmenu/settings/components.lua [hunk 2]
```diff
@@ -126,6 +126,40 @@ end, is_valid_number)
...
+			local formspec = setting_label..scrollbar_options..scrollbar_formspec
+
+			return formspec, 1.1
+		end,
+
+		on_submit = function(self, fields)
+			local raw_value = core.explode_scrollbar_event(fields[setting.name]).value
+			local value = (raw_value/steps) * (setting.max - setting.min) + setting.min
+			core.settings:set(setting.name, tostring(value))
+			return false
```
Comment (y5nw) [inline builtin/mainmenu/settings/components.lua:157]: The value should only be changed if the scrollbar is modified

```suggestion
			local event = core.explode_scrollbar_event(fields[setting.name])
			if event.type ~= "CHG" then
				return false
			end
			local raw_value = event.value
			local value = (raw_value/steps) * (setting.max - setting.min) + setting.min
			core.settings:set(setting.name, tostring(value))
			return true
```

File: builtin/mainmenu/settings/components.lua [hunk 3]
```diff
@@ -126,6 +126,40 @@ end, is_valid_number)
...
+
+			self.resettable = true
+
+			local setting_label = ("label[0,0.25;%s]"):format(
+				setting.readable_name)
+			local scrollbar_options = ("scrollbaroptions[min=0;max=%f;smallstep=1;largestep=10]"):format(
+				steps)
+			local scrollbar_formspec = ("scrollbar[0,0.5;%f,0.4;horizontal;%s;%f]"):format(
+				avail_w - 1, setting.name, scrollbar_value)
+			local formspec = setting_label..scrollbar_options..scrollbar_formspec
```
Comment (y5nw) [inline builtin/mainmenu/settings/components.lua:148]: Why not use [`table.concat`](https://www.lua.org/manual/5.1/manual.html#pdf-table.concat)? That would make sense if you want to add more (e.g. value display/input box) to this field.

**Comment by Athozus** [comment]
Your feature (which is great, imho) uses `scrollbar` formspec element : wouldn't it be better to create a global `slider` formspec element, which would also be usable in mods / other main menu tabs (or be used as a soft-coded solution for existing volume change slider) ?

**Comment by Zughy** [comment]
Feel free to reopen if you start working on this again

File: builtin/mainmenu/settings/components.lua (hunk 2 above)
Comment (mrkubax10) [reply builtin/mainmenu/settings/components.lua:157]: Reason why mouse dragging stops working is that after returning `true` GUI gets regenerated and this probably causes scrollbar to lose focus.

Comment (kotek900) [reply builtin/mainmenu/settings/components.lua:157]: normally regenerating the GUI doesn't make scrollbars loose focus, but for some minetest reason it does in the main menu

File: builtin/mainmenu/settings/components.lua (hunk 3 above)
Comment (mrkubax10) [reply builtin/mainmenu/settings/components.lua:148]: Rest of code in this file also does this so I guess it's better for consistency.

**Pull request closed without merging**